lib/proxy_api/external_ipam.rb

```diff
@@ -0,0 +1,278 @@
+      raise ::Foreman::Exception.new(N_(response['error'])) if response['error']
+      response['data']
+    rescue => e
+      raise ProxyException.new(url, e, N_("Unable to retrieve the next available IP for subnet %s from External IPAM."), subnet)
+    end
+
+    # Adds an IP address to the specified subnet in External IPAM. This will reserve the IP in the
+    # External IPAM database. If group is specified, the IP will be added to the subnet within the
+    # given group.
+    #
+    # Params: 1. ip:               IP address to be added
+    #         2. subnet:           The IPv4 or IPv6 subnet CIDR. (Examples: IPv4 - "100.10.10.0/24",
+    #                              IPv6 - "2001:db8:abcd:12::/124")
+    #         3. group(optional):  The name of the External IPAM group containing the subnet.
+    #
+    # Returns: true if IP was added successfully to External IPAM, otherwise false
+    #
+    # Responses from Proxy plugin:
+    #   Response if success:
+    #     {"code": 200}
+    #   Response if IP already reserved:
+    #     {"code": 409, "error": "IP address already exists"}
+    #   Response if subnet error:
+    #     {"code": 404, "error": "The specified subnet does not exist in External IPAM."}
+    #   Response if missing required params:
+    #     {"code": 400, "error": ["A 'cidr' parameter for the subnet must be provided(e.g. 100.10.10.0/24)","Missing 'ip' parameter. An IPv4 address must be provided(e.g. 100.10.10.22)"]}
+    #   Response if can't connect to External IPAM server
+    #     {"code": 500, "error": "Unable to connect to External IPAM server"}
+    def add_ip_to_subnet(ip, subnet, group = "")
+      raise "subnet cannot be nil" if subnet.nil?
+      response = parse post({}, "/subnet/#{subnet}/#{ip}?group=#{URI.escape(group.to_s)}")
+      raise ::Foreman::Exception.new(N_(response['error'])) if response['error']
+      response['code'] == 200
+    rescue => e
+      raise ProxyException.new(url, e, N_("Unable to add IP %s to the subnet %s in External IPAM."), ip, subnet)
+    end
+
+    # Get a list of groups from External IPAM. A group is analagous to a 'section' in phpIPAM, and
+    # is a logical grouping of subnets/ips.
+    #
+    # Params: None
+    #
+    # Returns: An array of groups on success, or a hash with a "message" key
+    #          containing error on failure.
+    #
+    # Responses from Proxy plugin:
+    #   Response if success:
+    #     {"code": 200, "data": [
+    #       {name":"Test Group","description": "A Test Group"},
+    #       {name":"Awesome Group","description": "A totally awesome Group"}
+    #     ]}
+    #   Response if no groups exist:
+    #     {"code": 200, "data": []}
+    #   Response if groups are not supported:
+    #     {"code": 501, "error": "Groups are not supported"}
+    #   Response if can't connect to External IPAM server
+    #     {"code": 500, "error": "Unable to connect to External IPAM server"}
+    def get_groups
+      parse get("/groups")
```

Comment:
Can you also add the raise when `error` is returned to align behaviour with  previous methods if possible?